---
title: "Data preprocessing for TRIPOD"
author: "Yuriko Harigaya, Nancy R. Zhang, Yuchao Jiang"
date: "`r format(Sys.Date())`"
output:
    html_document:
        highlight: pygments
        toc: true
vignette: >
  %\VignetteIndexEntry{preprocessing}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE, warning = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# 1 Introduction

This is a supplement to [the TRIPOD vignette](http://htmlpreview.github.io/?https://github.com/yharigaya/TRIPOD/blob/main/vignettes/TRIPOD.html).
Here, we demonstrate data preprocessing required for conducting
analysis using the TRIPOD package.
The input is sigle-cell ATAC/RNA multiomic data, which is publicly available from 
[10x Genomics](https://support.10xgenomics.com/single-cell-multiome-atac-gex/datasets). 
The data was generated in 4881 cells of the mouse embryonic brain at day 18 
and contains ATAC-seq and RNA-seq read counts for 144437 ATAC peaks 
and 32285 gene regions, respectively.

# 2 Required packages

The TRIPOD package is available from [GitHub](https://github.com/yharigaya/TRIPOD).

```{r install, eval = TRUE, message = FALSE, warning = FALSE}
if (!requireNamespace("devtools", quietly = TRUE)) install.packages("devtools")
devtools::install_github("yharigaya/TRIPOD/package")
```

```{r setup, eval = TRUE, message = FALSE}
library(TRIPODprep)
```

Additionally, the following packages are required to run this vignette.

```{r packages, cache = FALSE, message = FALSE}
library(Seurat)
library(Signac)
library(GenomicRanges)
library(dplyr)
library(ggplot2)
library(EnsDb.Mmusculus.v79)
library(GenomeInfoDb)
library(patchwork)
library(BiocParallel)
library(chromVAR)
library(JASPAR2020)
library(TFBSTools)
library(motifmatchr)
library(BSgenome.Mmusculus.UCSC.mm10)
```

# 3 Preprocessing
## 3.1 Creating a Seurat object

We first create a Seurat object essentially according to the tutorial [here](https://satijalab.org/seurat/articles/weighted_nearest_neighbor_analysis.html). 
The input data (`e18_mouse_brain_fresh_5k_filtered_feature_bc_matrix.h5`, `e18_mouse_brain_fresh_5k_atac_fragments.tsv.gz`, and `e18_mouse_brain_fresh_5k_atac_fragments.tsv.gz.tbi`) can be downloaded 
from the 10x Genomics website
[here](https://support.10xgenomics.com/single-cell-multiome-atac-gex/datasets).

```{r, include = FALSE, eval = TRUE, message = FALSE, warning = FALSE}
# for testing on a local machine only
file <- "e18_mouse_brain_fresh_5k_filtered_feature_bc_matrix.h5"
dir.data <- "~/Dropbox/TRIPOD/data"
path <- file.path(dir.data, file)
```

In the following code, `"/path/to/your/hdf5/file"` needs to be replaced with
the path to `e18_mouse_brain_fresh_5k_filtered_feature_bc_matrix.h5`.

```{r, eval = FALSE, message = FALSE, warning = FALSE}
path <- "/path/to/your/hdf5/file"
```

The following code reads in the hdf5 file from 10x Genomics, which contains
both RNA and ATAC-seq data.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
inputdata.10x <- Read10X_h5(path); rm(path)
```

```{r, eval = TRUE, message = FALSE, warning = FALSE}
# extract RNA and ATAC data
rna.counts <- inputdata.10x$`Gene Expression`
atac.counts <- inputdata.10x$Peaks
```

We first create a Seurat object using the RNA-seq data.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
# create a Seurat object
e18 <- CreateSeuratObject(counts = rna.counts)
e18[["percent.mt"]] <- PercentageFeatureSet(e18, pattern = "^mt-")
```

We then add the ATAC-seq data to the Seurat object. We focus on
ATAC peak regions on autosomes.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
# select peaks in standard chromosomes
grange.counts <- StringToGRanges(rownames(atac.counts), sep = c(":", "-"))
grange.use <- seqnames(grange.counts) %in% standardChromosomes(grange.counts)
atac.counts <- atac.counts[as.vector(grange.use), ]
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79)
seqlevelsStyle(annotations) <- "UCSC"
genome(annotations) <- "mm10"
```

In the following code, `"/path/to/your/atac/fragmenets/file"` needs to be replaced with
the path to `e18_mouse_brain_fresh_5k_atac_fragments.tsv.gz`.
The index file, `e18_mouse_brain_fresh_5k_atac_fragments.tsv.gz.tbi`, must be
in the same directory.

```{r, include = FALSE, eval = TRUE, message = FALSE, warning = FALSE}
# for testing on my local machine only
dir.data <- "~/Dropbox/TRIPOD/data"
file <- "e18_mouse_brain_fresh_5k_atac_fragments.tsv.gz"
path <- file.path(dir.data, file)
```

```{r, eval = FALSE, message = FALSE, warning = FALSE}
path <- "/path/to/your/atac/fragmenets/file"
```

```{r, eval = TRUE, message = FALSE, warning = FALSE}
chrom.assay <- CreateChromatinAssay(
  counts = atac.counts, sep = c(":", "-"),
  genome = "mm10", fragments = path,
  min.cells = 10, annotation = annotations
); rm(path)
e18[["ATAC"]] <- chrom.assay
```

## 3.2 Adding cell type annotations (optional)

Although TRIPOD does not strictly require cell type information, it can substantially enrich 
the analysis. In our analysis, we use cell type annotation data obtained by
transferring cell type labels from an independent scRNA-seq reference using
SAVERCAT (ref). We load the `label.severcat` data, 
which is included in the TRIPOD package, and add it to the Seurat object.
Alternatively, the `label.seurat` data, which was obtained using Seurat V3 (ref),
can be used. 

```{r eval = TRUE, message = FALSE, warning = FALSE}
data(label.savercat)
e18$label.savercat <- label.savercat
```

## 3.3 Dimension reduction and visualization

Cell clustering and visualization give an informative overview of the data properties.
For this purpose, we first perform pre-processing and dimensional reduction
on the ATAC and RNA modalities independently according to [the Seurat tutorial](https://satijalab.org/seurat/articles/weighted_nearest_neighbor_analysis.html).

The following code normalizes the RNA-seq data, performs principal component
analysis (PCA), and constructs a Uniform Manifold Approximation and Projection
(UMAP) graph.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
# perform RNA analysis
DefaultAssay(e18) <- "RNA"
e18 <- SCTransform(e18, verbose = FALSE)
e18 <- RunPCA(e18)
e18	<- RunUMAP(e18, reduction = "pca", dims = 1:50, 
	reduction.name = "umap.rna", reduction.key = "rnaUMAP_")
```

The following code normalizes the ATAC-seq data, performs partial singular
value decomposition, and constructs a UMAP graph.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
# perform ATAC analysis
# exclude the first dimension as this is typically correlated with sequencing depth
DefaultAssay(e18) <- "ATAC"
e18 <- RunTFIDF(e18)
e18 <- FindTopFeatures(e18, min.cutoff = "q0")
e18 <- RunSVD(e18)
e18 <- RunUMAP(e18, reduction = "lsi", dims = 2:50, 
	reduction.name = "umap.atac", reduction.key = "atacUMAP_")
```

Additionally, we calculate a weighted-nearest neighbor (WNN) graph, 
which represents a weighted combination of RNA and ATAC-seq modalities.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
# calculate a WNN graph
e18 <- FindMultiModalNeighbors(e18, reduction.list = list("pca", "lsi"), 
	dims.list = list(1:50, 2:50))
e18 <- RunUMAP(e18, nn.name = "weighted.nn", reduction.name = "wnn.umap", 
	reduction.key = "wnnUMAP_")
# e18 <- FindClusters(e18, graph.name = "wsnn", algorithm = 3, verbose = FALSE)
```

We then visualize the cell-type labels using color-coded UMAP plots 
based on RNA-seq data, ATAC-seq data, and the WNN analysis.

```{r, eval = TRUE, message = FALSE, warning = FALSE, fig.width = 9, fig.height = 4, fig.align = "center"}
p1 <- DimPlot(e18, reduction = "umap.rna",  group.by = "label.savercat", 
	label = TRUE, label.size = 2.5, repel = TRUE) +
	ggtitle("RNA")
p2 <- DimPlot(e18, reduction = "umap.atac",  group.by = "label.savercat", 
	label = TRUE, label.size = 2.5, repel = TRUE) +
	ggtitle("ATAC")
p3 <- DimPlot(e18, reduction = "wnn.umap", group.by = "label.savercat", 
	label = TRUE, label.size = 2.5, repel = TRUE) +
	ggtitle("WNN")
p1 + p2 + p3 & NoLegend() & theme(plot.title = element_text(hjust = 0.5))
```

## 3.4 Filtering cells based on cell labels

Cell type labels can be used to filter the single cell data, which can,
in turn, increase signal-to-noise ratio.
We note that this procedure is inevitably subjective and that it is advisable
to experiment with different filtering strategies.
We merely illustrate an example case below, in which our basic goals
are as follows.

* keep cells with biologically reasonable cell labels
* remove cells that belong to cell types with a small number of cells

We first examine the number of cells of each cell type.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
sort(table(e18$label.savercat), decreasing = TRUE)
```

Based on the above information as well as existing biological knowledge, we
focus on the seven cell types, "Cajal-Retzius",
"Cortical or hippocampal glutamatergic", "Forebrain GABAergic", "Glioblast",                         Neuroblast, "Oligodendrocyte", and "Radial glia".

```{r, eval = TRUE, message = FALSE, warning = FALSE}
# filter cells based on cell labels
celltypes.to.keep <- c("Cajal-Retzius", 
	"Cortical or hippocampal glutamatergic",
  "Forebrain GABAergic", "Glioblast",                            
  "Neuroblast", "Oligodendrocyte", "Radial glia") 
keep <- which(e18$label.savercat %in% celltypes.to.keep)
e18 <- subset(e18, cells = keep)
e18$celltype <- e18$label.savercat
e18$label.savercat <- NULL
```

For simplicity, we rename "Cortical or hippocampal glutamatergic" and
"Forebrain GABAergic" as "Glutamatergic" and "GABAergic", respectively.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
e18$celltype[e18$celltype == "Cortical or hippocampal glutamatergic"] <- "Glutamatergic"
e18$celltype[e18$celltype == "Forebrain GABAergic"] <- "GABAergic"
```

We convert the cell type data to factors and set the factor levels 
based on their developmental order 
("Radial glia", "Neuroblast", "GABAergic", "Glutamatergic", "Glioblast", "Oligodendrocyte", 
and "Cajal-Retzius").

```{r, eval = TRUE, message = FALSE, warning = FALSE}
levels.celltype  <- c(
	"Radial glia", "Neuroblast",
	"GABAergic", "Glutamatergic",
	"Glioblast", "Oligodendrocyte",
	"Cajal-Retzius"
)
e18$celltype <- factor(e18$celltype, levels = levels.celltype)
```

## 3.5 Filtering cells based on sequencing depths

We then remove cells with low sequencing coverage or those with high proportions
of mitochondrial reads, the latter of which is indicative of dying cells.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
e18 <- subset(
  x = e18,
  subset = nCount_ATAC < 1e5 &
    nCount_ATAC > 1e3 &
    nCount_RNA < 1e5 &
    nCount_RNA > 1e3 &
    percent.mt < 20
)
```

We then visualize the filtered data.

```{r, eval = TRUE, message = FALSE, warning = FALSE, fig.width = 9, fig.height = 4, fig.align = "center"}
p1 <- DimPlot(e18, reduction = "umap.rna", group.by = "celltype", 
	label = TRUE, label.size = 2.5, repel = TRUE) +
	ggtitle("RNA")
p2 <- DimPlot(e18, reduction = "umap.atac", group.by = "celltype",  
	label = TRUE, label.size = 2.5, repel = TRUE) +
	ggtitle("ATAC")
p3 <- DimPlot(e18, reduction = "wnn.umap", group.by = "celltype", 
	label = TRUE, label.size = 2.5, repel = TRUE) +
	ggtitle("WNN")
p1 + p2 + p3 & NoLegend() & theme(plot.title = element_text(hjust = 0.5))
```

## 3.6 Run chromVAR

We next add TF motif information to the Seurat object
according to the Signac tutorial [here](https://satijalab.org/signac/articles/motif_vignette.html).

The following code obtains a list of position frequency matrices (PFMs) from 
the JASPAR2020 database. 
We use 633 TF motifs in human (taxon ID: 9696) because motifs are highly conserved 
and the list of mouse TF motifs from the database contains only 107 entries. 
Alternatively, the `species` argument can be set to `"vertibrate"`.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
# create a motif object
pfm.set <- getMatrixSet(
	x = JASPAR2020,
	opts = list(species = 9606, all_versions = FALSE)) # human
```

In some cases, the TF motif information may not fully incorporate the existing
knowledge in the literature.
For example, in our analysis of the mouse embryonic brain data,
we replace the motif for the Olig2 TF with `CAGCTG`, which has been validated 
by previous genetic and biochemical studies (ref).

For this purpose, we first construct a PFM of the Olig2 motif.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
pfm.olig2 <- PFMatrix(
	ID = "PFM.Olig2",
	name = "Olig2",
	strand = "+",
	bg = c(A = 0.25, C = 0.25, G = 0.25, T = 0.25),
  tags = list(
  	species = "10090",
    tax_group = "vertebrates"),
    profileMatrix = matrix(
    	c(0L,  1L, 0L,  0L,  0L,  0L, # A
        1L, 0L,  0L, 1L,  0L,  0L, # C
        0L,  0L,  1L,  0L, 0L,  1L, # G
        0L,  0L,  0L,  0L,  1L, 0L), # T
        byrow = TRUE, nrow = 4,
        dimnames=list(c("A", "C", "G", "T")))
)
```

The following code removes the original PFM of the Olig2 motif from the PFM set
and adds the custom PFM.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
pfm.set <- pfm.set[-which(sapply(pfm.set, function(x) x@name == "OLIG2"))]
pfm.set$PFM.Olig2 <- pfm.olig2
```

The following code creates a motif matrix.

```{r createmotifmatrix, eval = TRUE, message = FALSE, warning = FALSE}
motif.matrix <- CreateMotifMatrix(
  features = granges(e18),
  pwm = pfm.set,
  genome = "mm10",
  use.counts = FALSE
)
```

Once an optimal set of PFM is obtained, it can be added to the Seurat object
using the following code.

```{r runchromvar, eval = TRUE, message = FALSE, warning = FALSE}
motif.object <- CreateMotifObject(data = motif.matrix, pwm = pfm.set)
e18 <- SetAssayData(e18, assay = "ATAC", slot = "motifs", 
	new.data = motif.object)
e18 <- RunChromVAR(
  object = e18,
  genome = BSgenome.Mmusculus.UCSC.mm10
)
```

## 3.7 Creating objects required for model fitting

We then create objects containing information that is necessary for model fitting, 
namely, `transcripts.gr`, `peaks.gr`, `motifxTF`, and `peakxmotif`. 
See the documentation for more detailed descriptions of these objects.

The `transcripts.gr` object is a GRanges object from [the GenomicRanges package](https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html)
and contains chromosomal coordinates of annotated protein containing genes.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
DefaultAssay(e18) <- "ATAC"
transcripts.gr <- Signac:::CollapseToLongestTranscript(ranges = Annotation(e18))
transcripts.gr <- transcripts.gr[transcripts.gr$gene_biotype == "protein_coding"]
transcripts.gr <- transcripts.gr[seqnames(transcripts.gr) %in% 
		paste("chr", 1:19, sep = "")]
transcripts.gr <- sort(transcripts.gr)
transcripts.gr
```

The `peak.gr` object is a GRanges object from the GenomicRanges package
and contains chromosomal coordinates of the ATAC peak regions in the multiomic
data.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
peaks.gr <- e18@assays$ATAC@ranges
```

The `motifxTF` object is a matrix containing mapping
between TFs and their binding motifs.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
motifxTF <- unlist(e18@assays$ATAC@motifs@motif.names)
motifxTF <- cbind(names(motifxTF), motifxTF)
colnames(motifxTF) <- c("motif", "TF")
```

The `peakxmotif` object is a binary sparse matrix containing indicators
as to whether a given motif is preset in a given ATAC peak region.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
peakxmotif <- e18@assays$ATAC@motifs@data 
peakxmotif[1:5, 1:5]
dim(peakxmotif)
```

We then convert the human gene symbols to mouse counterparts.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
motifxTF[, 2] <- stringr::str_to_title(tolower(motifxTF[, 2]))
```

We only keep motifs when genes encoding corresponding TFs are present in the
RNA-seq data.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
peakxmotif <- peakxmotif[, motifxTF[, 2] %in% rownames(e18@assays$RNA)]
motifxTF <- motifxTF[motifxTF[, 2] %in% rownames(e18@assays$RNA), ]
```

We note that our current implementation
excludes cases where a TF corresponds to multiple motifs or vice versa. 
We plan to provide an option to include such degenerate cases in a future version
of the package.

## 3.8 Further filering of the data

Since the normalization step (SCTranform) do not return values for all input,
we perform further data filtering
to ensure that the same set of genes is included in all objects. 

```{r, eval = TRUE, message = FALSE, warning = FALSE}
# take the intersection of the genes
genes <- intersect(transcripts.gr$gene_name, rownames(e18@assays$SCT))
DefaultAssay(e18) <- "RNA"
e18@assays$RNA <- subset(e18@assays$RNA, 
	features = match(genes, rownames(e18@assays$RNA)))
e18@assays$SCT <- subset(e18@assays$SCT, 
	features = match(genes, rownames(e18@assays$SCT)))
transcripts.gr <- transcripts.gr[match(genes, transcripts.gr$gene_name)]
peakxmotif <- peakxmotif[, motifxTF[, 2] %in% genes]
motifxTF <- motifxTF[motifxTF[, 2] %in% genes, ]
e18@assays$chromvar <- subset(
	e18@assays$chromvar,
	features = match(motifxTF[, 1], rownames(e18@assays$chromvar)))
```

We then perform normalization and dimension reduction again for the filtered data.

```{r renormalization, eval = TRUE, message = FALSE, warning = FALSE}
# perform re-normalization
# RNA analysis
DefaultAssay(e18) <- "RNA"
e18 <- SCTransform(e18, verbose = FALSE)
e18 <- RunPCA(e18)
e18	<- RunUMAP(e18, reduction = "pca", dims = 1:50, 
	reduction.name = "umap.rna", reduction.key = "rnaUMAP_")

# ATAC analysis
DefaultAssay(e18) <- "ATAC"
e18 <- RunTFIDF(e18)
e18 <- FindTopFeatures(e18, min.cutoff = "q0")
e18 <- RunSVD(e18)
e18 <- RunUMAP(e18, reduction = "lsi", dims = 2:50, 
	reduction.name = "umap.atac", reduction.key = "atacUMAP_")

# recalculate a WNN graph
e18 <- FindMultiModalNeighbors(e18, reduction.list = list("pca", "lsi"), 
	dims.list = list(1:50, 2:50))
e18 <- RunUMAP(e18, nn.name = "weighted.nn", 
	reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")
e18 <- FindClusters(e18, graph.name = "wsnn", 
  algorithm = 3, verbose = FALSE) # this may not be necessary
```

The following code visualizes the data on recalculated graphs.

```{r, eval = TRUE, message = FALSE, warning = FALSE, fig.width = 9, fig.height = 4, fig.align = "center"}
p1 <- DimPlot(e18, reduction = "umap.rna",  group.by = "celltype", 
	label = TRUE, label.size = 2.5, repel = TRUE) +
	ggtitle("RNA")
p2 <- DimPlot(e18, reduction = "umap.atac",  group.by = "celltype", 
	label = TRUE, label.size = 2.5, repel = TRUE) +
	ggtitle("ATAC")
p3 <- DimPlot(e18, reduction = "wnn.umap", group.by = "celltype", 
	label = TRUE, label.size = 2.5, repel = TRUE) +
	ggtitle("WNN")
p1 + p2 + p3 & NoLegend() & theme(plot.title = element_text(hjust = 0.5))
```

## 3.9 Partitioning single cells into metacells

To alleviate the issue of sparsity of the single cell data, we next partition the cells
into clusters, which we call metacells. The models in the TRIPOD package are
based on metacells, not single cells. Cells can be clustered based on either the ATAC
or RNA modality. It is also possible to use the weighted nearest neighbor (WNN) method (ref).
For the mouse embyonic barin data, we use the RNA modality (see above).

To obtain a set of metacells suited for model fitting, it is necessary to select
an optimal resolution of clustering to ensure the following.

* a sufficient number of cells per metacell for resolving the sparsity issue
* a sufficient number of metacells to allow for model fitting

Our empirical approach is to select a resolution that gives 80 or more metacells,
the majority of which contain 20 or more single cells. For this purpose, 
we first construct a shared nearest neighbor graph.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
# get rna meta-cells by clustering cells on basis of their scRNA-seq profiles
DefaultAssay(e18) <- "SCT"
e18 <- FindNeighbors(e18, reduction = "pca", dims = 1:30)
```

Based on the graph, we then perform clustering at several resolutions.
The following code generates a data frame containing the resolutions,
the number of resultant clusters, and the number of clusters with fewer than
threshold number of single cells. The threshold is set to 20 cells.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
num.clusters <- optimizeResolution(
	object = e18,
	graph.name = "SCT_snn",
	assay.name = "SCT",
	resolutions = seq(10, 35, 5),
	min.num = 20
)
num.clusters
```

Based on the above analysis, we select the resolution of 15.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
res <- 15
e18 <- FindClusters(e18, graph.name = "SCT_snn",
    	resolution = res, verbose = FALSE)
```

In our procedure, the factor levels need to be in an ascending numerical order.
The following code shows that the ordering is not desirable.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
levels(e18$SCT_snn_res.15)
```

We reorder the factor levels using the following code.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
tmp <- as.character(e18$SCT_snn_res.15)
levels.15 <- as.character(sort(as.numeric(levels(e18$SCT_snn_res.15))))
tmp <- factor(tmp, levels = levels.15)
e18$SCT_snn_res.15 <- tmp
levels(e18$SCT_snn_res.15)
```

We then overwrite
the `seurat_cluster` column with the `SCT_snn_res.15` column.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
e18@meta.data$seurat_clusters <- e18@meta.data$SCT_snn_res.15
e18@meta.data$SCT_snn_res.15 <- NULL
```

We now obtain matrices containing normalized RNA expression and chromatin accessibility per metacell using the `getMetacellMatrix` function. 

```{r, eval = TRUE, message = FALSE, warning = FALSE}
metacell.rna <- getMetacellMatrix(
	object = e18,
  cluster = "seurat_clusters",
	assay = "RNA")

metacell.peak <- getMetacellMatrix(
	object = e18,
  cluster = "seurat_clusters",
	assay = "ATAC")
```

We then remove metacell clusters with fewer than 20 cells from the matrices
as well as cells in the removed metacell clusters from the Seurat object.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
# remove metacell clusters with fewer than 20 cells
remove <- as.vector(table(e18$seurat_clusters)) < 20
names.remove <- names(table(e18$seurat_clusters))[remove]
metacell.rna <- metacell.rna[!remove, ]
metacell.peak <- metacell.peak[!remove, ]

# remove cells in the removed metacell clusters from the Seurat object
e18 <- e18[, !(e18$seurat_clusters %in% names.remove)]
e18$seurat_clusters <- droplevels(e18$seurat_clusters)
```

It is necessary to save the Seurat object (`e18`) for running [the TRIPOD vignette](http://htmlpreview.github.io/?https://github.com/yharigaya/TRIPOD/blob/main/vignettes/TRIPOD.html)
since the object is not included in the TRIPOD package.

```{r, eval = FALSE}
file <- "e18.rds"
path <- file.path("path/to/your/dir", file)
saveRDS(e18, path); rm(path)
```

## 3.10 Assigning cell types and colors for visualization

For visualization purposes, it is convenient to assign colors to cell types
and create objects storing mapping between them. For this purpose, 
we first assign cell types to metacells.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
metacell.celltype <- getCellTypeForMetacell(e18,
	celltype.col.name = "celltype", cluster.col.name = "seurat_clusters")
```

We then assign colors to cell types. The following code generates a mapping 
between cell types and colors for single cells.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
sc.color.map <- getColorsForSingleCells(
	object = e18, reduction = "umap.rna", celltype.col.name = "celltype")
head(sc.color.map)
```

The following code creates a mapping between cell types and colors for metacells.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
metacell.color.map <- getColorsForMetacells(
	metacell.celltype = metacell.celltype, sc.color.map = sc.color.map)
head(metacell.color.map)
```

It is also convenient to have a mapping between cell types and colors
themselves.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
color.map <- getColors(ordered.celltypes = levels(e18$celltype),
	metacell.color.map = metacell.color.map)
color.map
```

## 3.11 Getting a list of target genes

We typically run model fitting for top-ranked highly variable genes (HVGs)
identified by Seurat. The following code gives a ordered list of top 3000 HVGs.

```{r, eval = TRUE, message = FALSE, warning = FALSE}
# obtain highly variable genes based on SCT
DefaultAssay(e18) <- "SCT"
head(VariableFeatures(e18))
length(VariableFeatures(e18))
hvg <- VariableFeatures(e18)
```

# Session information

```{r}
sessionInfo()
```
